content/posts/flp-inside-out.md

```diff
@@ -0,0 +1,78 @@
+
+This part might be a bit confusiong unless we understand what is the set of processes taking steps in a schedule. Step made by process `p` is defined as a `receive(p, m)` (followed by internal state change and number of `send` operations). So the set of processes taking step in a schedule is a set of all processes `p` that make at least one receive operation in that schedule. It's important to realize that a single step defines not only a process that receives a message but also an exact message that gets received by that process. 
+
+With that in mind Lemma 1 becomes quite intuitive. An internal state of a process and messages that it has sent in a given schedule is fully determined by set of messages it received and order of that receival. If a schedule does not contain process `p` it cannot affect `p` in any way (either by chaning the message set or changing set of messages received).
+
+### Lemma 2
+
+#### Chain of initial configurations
+
+> Any two inital configurations are joint by a chain of initial configurations, each adjacent to the next. Heance, there must exists a 0-valent initial configuration C0 adjacent to a 1-valent initial configuration C1.
+
+This might not be obvious at the first sight. Lets start by look at the set of all inital configurations for 3 processes. An initial configuration is fully specified by values of 3 `x` registers each storing either `0` or `1`. There are 2^3 initial configurations that we can represent as 3-digit binary number. A chain of initial configurations in which one differs by a single bit is simply a enumarion of [Gray coding](TODO: add ref) for 3-digit binary number. In our case, `000` -> `001` -> `011` -> `010` -> `110` -> `101` -> `100`.
+
+If the chain contains both 1-valent and 0-valent intial configuration it's trivial to see that we can travers the sequence from one to the other and must arrive at an edge that connects 1-valent and 0-valent configurations that differ by a single bit i.e. by a single `x` register.
+
+//TODO: include a tesseract of 
+
+#### Internal state of the processes are not important
+
+But what about internal state of the process? Why isn't it mentioned in anywhere in the lemma? Does this imply that each process must start with the same internal state? Can we "break" the proof by coming up with an alogrithm that starts off with processes having different internal state? 
+
+The answer is ... we don't care if all processes have the same internal state or not. If an alogrithm is deterministic it has to define the internal state of each process before any of the values for `x` registers get defined. Futhermore, the state can depend only on some knowledge available before the algorithm starts e.g. process identifier. What this means is that even if each process has different internal state, these (different) states will be same for each possible permutation of the `x` register values if the algorithm is deterministic. 
```

Comment:
Nitpicking: comma before e.g.

Reply:
> if the algorithm is deterministic

Yeah, but if it's not? I guess it's worth mentioning that all algorithms are deterministic if we extract the randomness to the input register.

On the other hand, what if we had a truly randomized algorithm? Like based on a quantum computations which -  according to the Copenhagen interpretation - cannot be represented as a deterministic function f(x) with x encompassing the history or the world.